### DISCUSSION TITLE
Can I use a value object a Blazor route?

### DISCUSSION THREAD

**Discussion by MrYossu** [SteveDunn/Vogen#729] [Q&A] [answered]
Not strictly, a Vogen question, but I'd be surprised if no Vogen user had come across this before, so thought it would be a good place to ask.

I have a Blazor app, and one page shows the details of a company.

Before adding Vogen to the app, the first line of the component looked like this...

```c#
@page $"/company-{{Id:int}}"
```

The component had a corresponding parameter as follows...

```c#
  [Parameter]
  public int Id { get; set; }
```

That worked fine. I've now installed Vogen, and changed the type of the `Id` property on the `Company` model to be a value object...

```c#
[ValueObject]
public partial struct CompanyId;
```

I changed the `@page` directive on the Blazor component to use this...

```c#
@page $"/company-{{Id:CompanyId}}"
```

...and changed the type of the parameter to match, but this gives a run-time exception...

>InvalidOperationException: The constraint reference 'CompanyId' could not be resolved to a type. Register the constraint type with 'Microsoft.AspNetCore.Routing.RouteOptions.ConstraintMap'.

I tried adding the following...

```c#
builder.Services.Configure<RouteOptions>(opt => opt.ConstraintMap.Add("CompanyId", typeof(CompanyId)));
```

...which required me to change the value object to look like this...

```c#
[ValueObject]
public partial struct CompanyId : IParameterPolicy;
```

...but this gave a run-time exception...

>InvalidCastException: Unable to cast object of type 'System.String' to type 'VogenPlay.Common.Models.CompanyId'.

My guess is that I've got the wrong thing as the first parameter to `Add`, but I can't find any explanation as to what this is supposed to be.

Anyone able to help? Thanks

**Comment by SteveDunn** [comment]
I'm pretty sure it can be used in Blazor. I'm not able to test at the moment, but when I'm back online, I'll give this scenario a try (and add a project to the Samples folder).
Thanks for the question!

**Comment by viceroypenguin** [comment, marked as answer]
@MrYossu Unfortunately, Vogen ids cannot be used in Blazor routes. This is a Blazor limitation, in that it does not yet respect `IParseable<>` or any other mechanism for providing a parse function to the route system. This is a limitation I've found in my work with Vogen and Blazor as well.

**Reply by MrYossu** [reply]
@viceroypenguin I had a suspicion that this might be the answer. Shame, but I guess you can only go so far with this. 

I got around it by adding a read-only property to my component and using that in the code...

```c#
  [Parameter]
  public int Id { get; set; }

  private CompanyId CompanyId =>
    CompanyId.From(Id);
```

Thanks for the reply

**Reply by MrYossu** [reply]
@viceroypenguin Do you know if value objects can be used in Blazor component binding? I'm [struggling with this](https://github.com/SteveDunn/Vogen/discussions/731). Do you have any thoughts?
